client/blocks/jitm/index.jsx

```diff
@@ -0,0 +1,45 @@
+/**
+ * External Dependencies
+ */
+import React from 'react';
+import { connect } from 'react-redux';
+
+/**
+ * Internal dependencies
+ */
+import { getSelectedSite } from 'state/ui/selectors';
+import Banner from 'components/banner';
+
+const unescape = ( str ) => {
+	return str.replace( /&#(\d+);/g, ( match, entity ) => String.fromCharCode( entity ) );
+};
+
+const JITM = ( props ) => {
+	if ( props.data.length === 0 || ! props.currentSite ) {
+		return null;
+	}
+
+	const jitm = props.data[ 0 ];
+
+	return (
+		<Banner
+			callToAction={ unescape( jitm.CTA.message ) }
+			title={ unescape( jitm.content.message ) }
+			description={ unescape( jitm.content.description ) }
+			disableHref
+			dismissPreferenceName={ jitm.id }
+			dismissTemporary={ false }
+			event={ `jitm_nudge_click_${ jitm.id }` }
+			href={ `https://jetpack.com/redirect/?source=jitm-${ jitm.id }&site=${ props.currentSite.domain }` }
+		/>
+	);
+};
+
+const mapStateToProps = ( state ) => (
+	{
+		currentSite: getSelectedSite( state ),
+		data: state.jitm.jitms.data ? state.jitm.jitms.data : [],
```

Comment:
If `state.jitm` or `state.jitm.jitms` was ever `undefined` this would error. Maybe we should be more defensive and/or use `get` from lodash.

Reply:
A selector is probably better here, in general.

Reply:
As a general note this is very opaque from the component's perspective. What is `data` and what does it mean? What is the React component actually trying to render? If it's just a JITM object, then how about passing that in instead of some nebulously-wrapped `data` attribute?

Even better, what about passing in the simple data types that this component is rendering?

```js
<JITM { ...{ id, ctaMessage, message, description, domain } } />
```

At this point we have an incredibly easy component to test, one which isn't bound to the intricacies of the API, one which needs no escaping (can and should be done when the API response comes into Calypso - see `fromApi()` in the data layer), and one which is simple to "fake" into specific states.

Reply:
In this sense, `opaque` means "hidden or obscured" in the sense that we cannot see through an opaque wall. If the component can know little about data given to it or what it means then we're going to have trouble when that data unexpectedly changes.

As a general rule we have decided in Calypso to secure our interfaces via selectors to allow for flexibility in the underlying storage and retrieval of data out of state. We have a global list of selectors and a search interface built to find them inside of the Calypso Dev Docs.